repose-aggregator/core/repose-core/src/main/java/org/openrepose/core/services/reporting/metrics/MetricsJmxObjectNameFactory.java

```diff
@@ -23,12 +23,10 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import javax.inject.Named;
 import javax.management.MalformedObjectNameException;
 import javax.management.ObjectName;
 import java.util.Hashtable;
 
-@Named
```

Comment:
This was not actually being wired in anywhere, so to prevent confusion, I just removed the annotation.

I actually considered leaving this a bean and wiring it in places, but we have some tests that don't like that approach. That is, this class is necessary for the `ReposeJmxNamingStrategy` bean to work. Right now, that bean instantiates this class itself so that it can always use an instance. If I wired this class in, it would break unit tests that do not scan the core classpath, but instead scan a tests classpath.

If we want this to be a bean, we can always make it so with some work in the future.

Reply:
So I went ahead and made this a bean, but there are a couple of things that I'm not super happy about.

For one, the interface this class implements comes from the metrics library. As a result, the metrics library becomes a direct dependency for core, where the `ReposeJmxNamingStrategy` lives. Of course, since the metrics service is currently a core service, we have already declared the metrics library as a core dependency... but this change couples things a bit more tightly. That may never matter, depending on what we do with the metrics service, but it still doesn't thrill me. Alternatively, I could wire in `MetricsJmxObjectNameFactory` explicitly instead of the `ObjectNameFactory` interface, which keeps the dependency on the service rather than the backing library. Considering the direction I went with the tests, would that be acceptable?

Two, I don't love the repeated code in the tests. That is, the classes which satisfy the bean dependency when we are scanning other packages. Of course, I don't really know of a way around that unless we modify the tests themselves or take a different approach.

It basically seems like we have three options here:

1. Don't make this class a bean. Either make it a singleton, or instantiate it when needed.
1. Make this class a bean that is defined with Java code and registered directly with the Spring context. This gets us around the package scanning in tests issue while keeping this a Spring bean without requiring us to stub implementations. Also, since the naming strategy bean itself is defined this way, there is some consistency there.
1. Make this class an annotated bean.

Reply:
So this opens a lot of questions about about how we want to gather metrics, reports metrics, and expose metrics. And who is responsible for what there. What do we want the metrics service to do? At the moment, I'm leaning towards metrics gathering being an intrinsic things repose does meaning that it's not the responsibility of the metrics service, but that it becomes a dependency more akin to log4j. This probably means we expose the registry as a bean that anything can wire in (assuming that the registry itself doesn't have the same kind of awareness that log4j/slf4j does). Then the metrics service is responsible for the other two needs reporting and exposure. In this case reporting is the tying in and dumping to graphite, and exposure is part of the non-existent admin api providing a restful interface into the metrics so people can potentially look in on repose live. I don't remember where we ended up landing with the metrics service when we hacked at it a bunch several months back, and what level of compatibility this has with that. I don't even know if this is a good way of looking at it. I think this way helps relieve some of the dependency weirdness you are pointing out here though? Did any of that make sense? Thoughts?

Reply:
I think that made sense, and that seems legit to me. Basically, we just need to register the `MetricRegistry` as a bean, which is pretty easy to do. In fact, I think the work we did a few months back actually makes that easier, since we stopped forcing all interactions with the registry to go through the service (now, you can get a reference to the registry from the service, which is similar to making it a bean, just a bit more work). 

Of course, the metrics service would have to acquire the `MetricRegistry` bean to be able to configure `Reporter`s to report the metrics to JMX and Graphite (and wherever else we may want to report metrics). At this point, I should note that anyone with a reference to the `MetricRegistry` will be able to create a reporter, which is pretty neat. I mean, that's already the case since the registry is exposed through the service, but still.

The only other thing to consider is the custom support we added for combining metrics. So the `SummingMeterFactory` and `MultiMeter`. These will still work regardless, and we can adjust the service to continue making it easy to access them, or we can move them into common utilities (since they are utilities for managing metrics in the registry rather than part of the service, really).

Oh, and there's the `MetricsJmxObjectNameFactory` that was the root of this conversation. Well, that's easy. Since the metrics library must be part of core, there is no problem using the interface. I would probably move this class to utilities as well. I was actually already considering doing that since this logic is now used for both Spring JMX beans and metrics JMX beans. I had hoped that there was a Java management interface for an `ObjectName` factory that I could use to make it a more standard utility, but unfortunately, there does not seem to be. So the interface was the only reason I kept it in the metrics service.

Reply:
To be clear I'm not saying to do these things now, but for this to be the direction we push things toward. Hopefully @Mario-Lopez and @wdschei chime in with their thoughts.

Reply:
Indeed.

As for the code here, I reverted the bean changes and did what made the most sense to me. I left the bean changes commit so that we have the history, and you can see both approaches. My new approach was to make the JMX `ObjectName` factory a common utility, since it really is our approach to naming any JMX bean at the moment, not just those reported as metrics. When I did that, I made the `ObjectName` creation method static, so there was no need to make a bean that had to be wired in places. To satisfy the contract for the metrics library, I kept `MetricsJmxObjectNameFactory` around as a very simple singleton wrapper to the common utility.

Reply:
I like the idea of exposing the registry as a bean so that anything that wants to report metrics can. I also like the ideas of making a facility for dumping to an external analysis tool as well providing an API for direct live queries. I also agree this is way out of scope of this task.